solarforecastarbiter/reference_forecasts/persistence.py

```diff
@@ -0,0 +1,344 @@
+
+    Parameters
+    ----------
+    observation : datamodel.Observation
+    data_start : pd.Timestamp
+        Observation data start. Forecast is inclusive of this instant if
+        observation.interval_label is *beginning* or *instantaneous*.
+    data_end : pd.Timestamp
+        Observation data end. Forecast is inclusive of this instant if
+        observation.interval_label is *ending* or *instantaneous*.
+    forecast_start : pd.Timestamp
+        Forecast start. Forecast is inclusive of this instant if
+        interval_label is *beginning* or *instantaneous*.
+    interval_length : pd.Timedelta
+        Forecast interval length
+    interval_label : str
+        instantaneous, beginning, or ending
+    load_data : function, default solarforecastarbiter.io.utils.load_data
+        A function that loads the observation data. Must have the same
+        signature has the default function.
+
+    Returns
+    -------
+    forecast : pd.Series
+        The persistence forecast.
+    """
+    # determine forecast end from forecast start and obs length
+    # assumes that load_data will return NaNs if data is missing.
+    forecast_end = forecast_start + (data_end - data_start)
+
+    # ensure that we're using times rounded to multiple of interval_length
+    _check_intervals_times(observation.interval_label, data_start, data_end,
+                           forecast_start, forecast_end, interval_length)
+
+    # get the data
+    load_data = _load_data if load_data is None else load_data
+    obs = load_data(observation, data_start, data_end)
+
+    # average data within bins of length interval_length
+    persistence_quantity = obs.resample(interval_length).mean()
+
+    # Make the forecast time index.
+    closed = datamodel.CLOSED_MAPPING[interval_label]
+    fx_index = pd.date_range(start=forecast_start, end=forecast_end,
+                             freq=interval_length, closed=closed)
+
+    # Construct the returned series.
+    # Use values to strip the time information from resampled obs.
+    # Raises ValueError if len(persistence_quantity) != len(fx_index), but
+    # that should never happen given the way we're calculating things here.
+    fx = pd.Series(persistence_quantity.values, index=fx_index)
+    return fx
+
+
+def persistence_scalar_index(observation, data_start, data_end, forecast_start,
+                             forecast_end, interval_length, interval_label,
+                             load_data=None):
+    r"""
+    Calculate a persistence forecast using the mean value of the
+    *observation* clear sky index or AC power index from *data_start* to
+    *data_end*.
+
+    In the example below, we use GHI to be concrete but the concept also
+    applies to AC power. The persistence forecast is:
+
+    .. math::
+
+       GHI_{t_f} = \frac{\overline{GHI_{t_{start}} \ldots GHI_{t_{end}}}}{\overline{GHI_{{clear}_{t_{start}}} \ldots GHI_{{clear}_{t_{end}}}}} \times GHI_{{clear}_{t_f}}
```

Comment:
I think ratio of the averages correctly describes the implementation for generic inputs. I believe the average of the ratios requires assertions about interval alignment that is inconsistent with the present implementation.

Reply:
Agree that the comment describes the implementation, but I disagree that it is what should be implemented.  The ratio of averages is not the average clear sky index which I thought we intended by a persistence index forecast.  Since the clear sky calculation is under our control, what is the barrier to implementation?

Reply:
No barrier, just thinking through it. I think we need to change

```python
pers_index = obs.mean() / clear_ref.mean()
```

to 

```python
clear_ref_resampled = clear_ref.resample(observation.interval_length, closed=closed, label=closed).mean()
pers_index = (obs / clear_ref_resampled).mean()
```

We probably want to ensure that the observation interval is divisible by the `clear_ref` interval. Perhaps choose `freq = pd.Timedelta('1min')` for simplicity. Otherwise we need to call interpolate, ffill, and bfill on clear_ref_resampled or accept NaNs for some fractions. Or we could find the largest factor of the observation interval length that is less than some threshold (currently 5 minutes). Neither of those is easy to explain.

Reply:
Ah, I see the wrinkle - even with average of ratios, at each interval we still compute a ratio of averages (average of data / average of clear sky) for the interval.  I think that's OK because the common understanding of irradiance is a time average over a short interval.  Same for power.  

"Otherwise …" agree these considerations are not worth the effort.

Reply:
My proposed implementation is not strictly correct for observations that are "instantaneous" (instrument sample time << interval length) and have interval length > 1 min. Say GHI is recorded for 10 seconds every 60 minutes. We want to calculate clear sky at each 60 minute point rather than the 60 minute interval average. We can handle observations that are either interval averages or instantaneous with:

```python
if observation.interval_label == 'instantaneous':
    freq = observation.interval_length
else:
    freq = pd.Timedelta('1min')
```